### DISCUSSION TITLE
2023-06-9 - Data Engineering #2 - Databases - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by lukecookssw** [SSWConsulting/SSW.Rules.Content#5804] [Chewing the Fat]
Hey data engineers!

Let's talk about [this rule](https://www.ssw.com.au/rules/use-the-right-database/)

**Comment by brydeno** [comment]
One reason to choose NoSql can be cost and simplicity. Cosmos Db in particularly gives you super simple geo replication with a single checkbox... But do some research Cosmos and NoSql in general have their perfect use cases...

**Comment by ncn-ssw** [comment]
Some programmers use NoSQL databases as an excuse to avoid data modeling. They perceive NoSQL's flexible schema and document-oriented nature as an opportunity to forego the structured approach of data modeling, believing that they can simply store data in a loosely structured manner. However, this mindset can lead to a lack of consistency and integrity in the data, making it difficult to query and analyze efficiently. 🤖⭐

**Comment by leahy268** [comment]
I've used lot's of SQL but never really used NoSQL.  Looks interesting but I fear that the lack of structure might make it harder to find the data that you need.

**Comment by tkapa** [comment]
I've used a lot of SQL. I think that the rule is very informative, but has the same issue of being a huge wall of text as the last one.

**Comment by AlexRothwell** [comment]
I've done plenty of work with SQL on various levels of abstraction and am keen to use NoSQL sometime

**Comment by jakebayliss** [comment]
As with a lot of other devs, I have done a lot of SQL but not a lot of NoSQL. Maybe NoSQL has much more specific use cases...

**Comment by ozdemir-mehmet** [comment]
Looking forward to more dev focussed topics

**Comment by vladislav-kir** [comment]
This was a great summary of differences between SQL and NoSQL

**Comment by matt-goldman** [comment]
I really like the flexibility of combining JSON columns with a SQL database. Although PostgreSQL supported this long before MS SQL did.

**Comment by bahjatalaadelssw** [comment]
This week's subject was a little too technical for me, but I'm looking forward to seeing new structures for databases in the future

**Comment by cvionnet** [comment]
Like others devs, I played quite a lot with TSQL but not so much with NoSQL (only with CosmosDB). I learnt about storing JSON data in SQL Server.

**Comment by chrisschultzssw** [comment]
I have limited experience with databases, and only with SQL (and some MySQL). It was interesting to learn more about NoSQL options.

**Comment by jeoffreyfischer** [comment]
I've only heard of SQL before so that was a good learning item!

**Comment by sethdaily** [comment]
I learned what both SQL and NoSQL mean. It was helpful because I have heard many a discussion about SQL but never been familiar with what it actually does.

**Comment by christoment** [comment]
The JSON in a table works quite good in PostgreSQL before, so keen to see how the MSSQL version works!

**Comment by AntPolkanov** [comment]
Like the rules! 

I've got some experience with SQL Server, DynamoDB, Realm and MongoDB. One of the main disappointments for me when I started learning NoSQL was the tight coupling between the way you store your data and your ability to query it. No wonder they are fast! In order to properly design a NoSQL database you need to have a clear understanding of the use cases.

**Comment by SamBWagner** [comment]
I've only really used SQL databases professionally and of course the general JSON data text file for random home files haha.

I'm keen to test out things like MongoDB for home projects or otherwise. I just tend not to do projects that require data. (making little games 'n' what not)

**Comment by UlyssesMaclaren** [comment]
SQL database -> data warehouse
noSQL database -> data lake

**Comment by JackDevAU** [comment]
I've used a fair amount of SQL and NoSQL databases, though mainly SQL since I've been at SSW. In regards to the rule, I think it could be improved. Lets make it have less text and a nice table comparing them.

**Comment by tombui99** [comment]
NoSQL can be easier to use and setup but I still much prefer SQL because of the structure and robustness. From my personal experience, SQL also has some amazing abilities when it comes to querying data unlike NoSQL. Furthermore, recently, SQL now has the JSON capabilities so I suspect the use of document database will get much lower.

**Comment by Levijj22** [comment]
Interesting topic.
I've learnt that when it comes to databases, SQL offers structure and consistency, while NoSQL provides flexibility and scalability. As someone in sales, i feel it's important to understand why a dev has chosen a particular database.

**Comment by Ben0189** [comment]
I have some experience in SQL, but haven't done any NoSQL

**Comment by jackreimers** [comment]
I've not used a NoSQL database before, but already had a basic understanding of them. It was interesting to learn more about them but I think it would be helpful to see one in action solving a problem that a SQL database can't to fully understand their benefits

**Comment by Abf09** [comment]
I have used SQL before mostly for uni, keen to try NoSQL

**Comment by Harry-Ross** [comment]
I've spent a fair amount of time with MongoDB (NoSQL) for personal projects, and found it to be very easy to use - however I sometimes encountered a couple pitfalls in comparison to structured data. I would love to greater understand the benefits/drawbacks of using a NoSQL vs SQL database.

**Comment by stephanfakossw** [comment]
Great summary, great haircut 😊

**Comment by Marxoz** [comment]
I am loving the focus on. data and databases. it is such an important concept to understand, whether you are working with the core database or just interacting with it.

**Comment by PennyWalker** [comment]
I enjoyed learning the difference between the two types of databases, I think this is really valuable :)

**Comment by Jord-Gui** [comment]
I've worked with SQL before, but not with NoSql

**Comment by amankumarrr** [comment]
I have extensive experience writing SQL queries, so I find it challenging to work with NoSQL databases when it comes to complex queries. The absence of joins in NoSQL contradicts the relational database concepts I'm familiar with. However, I'm keen to learn it.

**Comment by leehawkinsssw** [comment]
I was very experienced in the early days of NoSQL databases when Quest were building a tool called Toad for Cloud Databases to make them easier to access and manage for SQL-familiar developers. That product was before its time and didn't survive, sadly.

Now that NoSQL databases have become so ubiquitous, this CTF and rule are valuable for those who might not have encountered anything other than the well-known relational databases before.

**Comment by MattParkerDev** [comment]
I've heard a lot about NoSQL databases, but I haven't had the chance to work with them extensively. It seems like they offer more flexibility and scalability compared to traditional SQL databases. However, I'm also aware that NoSQL databases can be more challenging when it comes to complex querying and maintaining data integrity. Overall, I believe NoSQL has its perfect use cases, and it would be interesting to explore them further.

**Comment by zacharykeeping** [comment]
Good rule - breaks down the differences between SQL and NoSQL well.

**Comment by steven0x51** [comment]
Good rule that explains the different use cases for SQL and NoSQL. I have SQL experience but haven't done NoSQL yet, very keen to try NoSQL.

**Comment by bradystroud** [comment]
It seems like the future could be in hybrid solutions - SomeSQL™

**Reply by jackreimers** [reply]
How can I invest?

**Reply by bradystroud** [reply]
@jackreimers dm me

**Comment by tomek-i** [comment]
Worked dominantly with relational databases and MSSQL. For prototyping NoSQL databases are handy as you can focus on development and think about data second especially when there are frequent changes in schema and model due to experimentation or figuring out what data you going to record etc. 

Since most relational databases use pretty much the same syntax and SQL, NoSQL comes with new challenges by learning a whole new syntax. 

Having a user's table and figuring out the gender distribution between `male` / `female` is fairly straightforward in SQL but using MongoDB its somewhat complicated :- D
```
let grouped = await UserModel.aggregate([
    {
      $project: {
        male: {
          $cond: [
            {
              $eq: ["$gender", "male"],
            },
            {
              _id: "$_id",
              username: "$username",
              gender: "$gender",
            },
            "$ex",
          ],
        },
        female: {
          $cond: [
            {
              $eq: ["$gender", "female"], 
            },
            {
              _id: "$_id",
              username: "$username",
            },
            "$ex",
          ],
        },
      },
    },
    {
      $group: {
        _id: null,
        male: {
          $push: "$male",
        },
        female: {
          $push: "$female",
        },
        total: {
          $sum: 1,
        },
      },
    },
  ]).exec();
```

**Comment by lukecookssw** [comment]
In writing this week's CTF, I learned that a lot of benefits I thought were only available to SQL/noSQL are actually available in both, which surprised me. I haven't worked with noSQL in any meaningful fashion, and coming from almost 20 years in the SQL world it's quite difficult to break out of the relational DB school of thought.

**Comment by jomelsontanco** [comment]
Being an SQL guy primarily working with PLSQL (Oracle) during my career, and then dabbling with mssql (Microsoft) recently as an intern here in SSW, I initially found the term noSQL preposterous. This short tutorial opened my mind a little bit. noSQL actually has valid use cases!

**Comment by Dhruv-0987** [comment]
I think we need to keep doing hands on practice with SQL because LINQ is always easy to use and nobody really knows what's the actual query statement. Also we should consider using mongodb for large scale enterprise applications would make sense to handle more traffic with less latency

**Comment by AshAnil** [comment]
Nice rule - Now I know the difference between SQL and NOSQL 🤩

**Comment by jernejk** [comment]
As an EF Core expert, I prefer SQL over NoSQL, but support for NoSQL is getting better with each version.
It's also that most clients I have, didn't need scale and whatever scheme flexibility they needed, a simple JSON column solved that.

**Comment by william-liebenberg** [comment]
The rule is very wordy. A decision tree like this one from microsoft would be a good starting point to simplify the rule: https://learn.microsoft.com/en-us/azure/architecture/guide/technology-choices/data-store-decision-tree#select-a-candidate

More criteria to consider when selecting an appropriate data store:
https://learn.microsoft.com/en-us/azure/architecture/data-guide/technology-choices/analytical-data-stores

**Comment by duanxinhuan** [comment]
I love and hate no-sql a loooot!!!!!!!!!!!!!!

I remeber when I work on firebase, it was alll happy at the beginning but end up a nightmare......

That's when I learn that a traditional DB & sql are proficient and good!!!




For vector DB: I love it, I am keep to see how efficient it can be with mordern NLP techniques


I am not happy they just use cosine-similarities for vector db, I want something more!

I want them to use Tranformers!!!!!!!!

**Comment by jimmidier** [comment]
NoSQL is an essential skill for data engineers. For more details on NoSQL database please see the comment below

**Comment by sylhuang** [comment]
Recently I've been forced to use a No SQL db for one of my personal project. And I ended up splitting up collections just like how I would design a relational database 😅 (e.g. User collection + Visit collection with foreign keys, and I assume alternative is embedding user visits into user as an array). Wonder when it comes to db/schema design, what should have been done differently to better utilize the power of No SQL db?

**Comment by JaneYangssw** [comment]
Great rule! I never used No SQL in the past😁. I'll try to know more about it in the future.

**Comment by JerryLuo5799** [comment]
Great rule, NoSQL is awesome, I have used it in log, document, search and other scenarios.

**Comment by johnxu16** [comment]
Great video for introduces differences between SQL and NoSQL. NoSQL is too general to discuss since they all have their own use cases.

**Comment by AttackOnMorty** [comment]
I haven't used NoSQL in real-world scenarios. Keen to learn :)

**Comment by Aibono1225** [comment]
I have a lot to learn about databases, it's happy to explore the distinctions between SQL and NoSQL. :)

**Comment by tino-liu** [comment]
I remember when I first touched MongoDb after using SQL for awhile.. it definitely threw me off. Now that I'm back learning, and getting used to SQL again, I think it's about time i get thrown off again! But this time fully prepared.

**Comment by PothieuG** [comment]
I have never worked professionally with SQL, but I would really love to!

**Comment by Freego1783** [comment]
I mainly used MongoDB and I have a lot to learn about SQL. Keen to do so !

**Comment by ozair-ashfaque** [comment]
Worked with SQL but never used NoSQL. I would love to learn about it.

**Comment by drwharris** [comment]
Dont get sucked into using SQL Server JSON types in place of proper document storage.

**Comment by BrookJeynes** [comment]
Done a bit of both (SQL and NoSQL) in uni, both seem to have their benefits and downsides. Knowing of both though can help you pick the right tool for the job

**Comment by danielmackay** [comment]
The 'agility' provided by No-SQL databases, can come back to bite you when an application grows and you're not careful about managing the DB schema from the application.  I haven't used CosmosDB much, but I also hear the Azure costs can get out of control if you're not careful.

**Comment by TobyChurches** [comment]
I've mainly used SQL databases and touched on MongoDB with a University project. I still feel that I have a lot to learn from both

**Comment by wicksipedia** [comment]
Choosing between SQL and NoSQL is like the difference between building a LEGO set with a detailed instruction manual versus creating your own unique structure with a random pile of LEGO bricks.

If you have well-defined instructions and a specific end product in mind, SQL is your LEGO kit with a manual. Everything has its place, your pieces (or data) fit together in a prescribed, orderly manner, and the blueprint is clear.

On the flip side, if your project requires flexibility, creativity, and the ability to change and adapt, NoSQL is like a random pile of LEGO bricks. You're not confined by a set structure, you can add or remove pieces on the fly, and your final creation can be as unique and dynamic as your data.

But remember: just because you're holding a LEGO brick doesn't mean everything should become part of your creative masterpiece, and just because you have a detailed instruction manual doesn't mean every scenario requires one.

**Comment by bahjatalaadelssw** [comment]
[https://outlook-1.cdn.office.net/assets/reaction/like.png]     Bahjat Alaadel [SSW] reacted to your message:



[cid:ssw-circle(1)_4c7259a6-23f6-49e5-b933-230ca304562e.png]

Bahjat Alaadel
SSW Multimedia Specialist/Videographer
Stalk: SSW People<https://www.ssw.com.au/people/Bahjat-Alaadel?utm_source=emailsignature>   |   Walk: SSW Sydney<https://www.ssw.com.au/ssw/Company/Offices/?utm_source=emailsignature>   |   Talk: +61 451 082 968<https://www.ssw.com.au/ssw/Company/ContactUs.aspx>
[cid:youtube-icon_b8b5cda9-a973-4973-8763-cf3d4922e060.png] <https://www.youtube.com/c/SSWTV>    [cid:linkedin-icon_b153ac9a-eb66-47d9-9c51-ad20b42b8a0a.png]  <https://www.linkedin.com/company/ssw/>    [cid:twitter-icon_16ddd2de-0a0c-4b56-9986-6c5968e517d2.png]  <https://twitter.com/SSW_TV>    [cid:teams-icon_552803c2-9c44-4267-b563-7fe1a7a0e589.png]  ***@***.***>


________________________________

**Comment by JeanThirion** [comment]
I like Wicksy's LEGO metaphore, I think it works quite well. Thanks Dan for the repo - haven't used it yet, but surely will refer to it in the future.

I like this Data series, very important topic

**Comment by Hona** [comment]
I've used postgres & MartenDB for NoSQL & it worked very well.

https://martendb.io/
